File: benchmarks/cpp/heapsort.cpp
#include <cstdio>
#include <vector>
void sift_down(std::vector<long long>& a, int i, int n) {
    while (true) {
        int lg=i,l=2*i+1,r=2*i+2;
        if (l<n&&a[l]>a[lg]) lg=l;
        if (r<n&&a[r]>a[lg]) lg=r;
        if (lg==i) break;
        std::swap(a[i],a[lg]); i=lg;
    }
}
int main() {
    long long seed=55555; std::vector<long long> lst(100000);
    for (auto& v:lst){seed=(seed*1664525+1013904223)%4294967296LL;v=seed%100000;}
    int n=lst.size();
    for(int i=n/2-1;i>=0;i--) sift_down(lst,i,n);
    for(int end=n-1;end>0;end--){std::swap(lst[0],lst[end]);sift_down(lst,0,end);}
    printf("%lld\n", lst[50000]);
}

File: benchmarks/heapsort.py
def sift_down(a, i, n):
    while True:
        largest = i; l = 2*i+1; r = 2*i+2
        if l < n and a[l] > a[largest]: largest = l
        if r < n and a[r] > a[largest]: largest = r
        if largest == i: break
        a[i], a[largest] = a[largest], a[i]; i = largest
def heapsort(a):
    n = len(a)
    for i in range(n//2-1, -1, -1): sift_down(a, i, n)
    for end in range(n-1, 0, -1):
        a[0], a[end] = a[end], a[0]; sift_down(a, 0, end)
seed = 55555; lst = []
for _ in range(100000):
    seed = (seed * 1664525 + 1013904223) % 4294967296
    lst.append(seed % 100000)
heapsort(lst)
print(lst[50000])
